File: code/java/Generics.java
public class Generics {
// BEGIN_CODE
	public static class GenericClass<T> {
		private T t;
		
		public GenericClass(T t) {
			this.t = t;
		}
	}

	public static <T> T do_nothing(T t) {
		return t;
	}

	public static void main(String[] args) {
		GenericClass<Integer> gc_int
			= new GenericClass<>(2);
		GenericClass<Float> gc_float
			= new GenericClass<>(2.3f);

		int r1 = do_nothing(2);
		float r2 = do_nothing(2.3f);
	}
// END_CODE
}

File: code/cpp/generics.cpp
// BEGIN_CODE
template<typename T>
class generic_class {
	T t;
public:
	generic_class(const T &t)
		: t(t) { }
};

template<typename T>
T do_nothing(const T &t) {
	return t;
}

int main(int argc, char **argv) {
	generic_class<int> gc_int(2);
	generic_class<float> gc_float(2.3);

	int r1 = do_nothing(2);
	float r2 = do_nothing(2.3);

	return 0;
}
// END_CODE
